File: calco_production/page/plant_production_dashboard/plant_production_dashboard.py
from __future__ import annotations

from collections import OrderedDict

import frappe
from frappe.utils import flt, getdate, nowdate

from calco_erp.machine_setup import (
    PLANT_MACHINES,
    get_output_by_machine_rows,
    get_output_by_operator_rows,
    get_output_by_shift_rows,
)


@frappe.whitelist()
def get_dashboard_data(report_date: str | None = None) -> dict[str, object]:
    report_date = resolve_report_date(report_date)

    daily_production = get_daily_production_by_fg(report_date)
    extruder_utilization = get_extruder_utilization(report_date)
    output_by_machine = get_output_by_machine_rows(report_date)
    output_by_operator = get_output_by_operator_rows(report_date)
    output_by_shift = get_output_by_shift_rows(report_date)
    cost_by_product = get_cost_by_product(report_date)
    profit_by_product = get_profit_by_product(report_date, cost_by_product)
    rm_consumption = get_rm_consumption(report_date)
    qc_failures = get_qc_failures(report_date)
    dispatch_quantity = get_dispatch_quantity(report_date)

    return {
        "report_date": str(report_date),
        "cards": {
            "production_qty": round(sum(row["qty"] for row in daily_production), 3),
            "dispatch_qty": round(sum(row["qty"] for row in dispatch_quantity), 3),
            "rm_consumption_qty": round(sum(row["qty"] for row in rm_consumption), 3),
            "qc_failure_count": int(sum(row["count"] for row in qc_failures)),
            "machine_count": int(len(output_by_machine)),
            "operator_count": int(len(output_by_operator)),
            "shift_count": int(len(output_by_shift)),
            "avg_rm_cost_per_kg": round(
                sum(row["rm_cost_per_kg"] for row in cost_by_product) / len(cost_by_product), 4
            )
            if cost_by_product
            else 0,
            "avg_manufacturing_cost_per_kg": round(
                sum(row["manufacturing_cost_per_kg"] for row in cost_by_product) / len(cost_by_product), 4
            )
            if cost_by_product
            else 0,
            "gross_profit_total": round(sum(row["gross_profit_total"] for row in profit_by_product), 2),
            "avg_extruder_utilization": round(
                sum(row["utilization_pct"] for row in extruder_utilization) / len(extruder_utilization), 2
            )
            if extruder_utilization
            else 0,
        },
        "daily_production_by_fg": daily_production,
        "extruder_utilization": extruder_utilization,
        "output_by_machine": output_by_machine,
        "output_by_operator": output_by_operator,
        "output_by_shift": output_by_shift,
        "cost_by_product": cost_by_product,
        "profit_by_product": profit_by_product,
        "rm_consumption": rm_consumption,
        "qc_failures": qc_failures,
        "dispatch_quantity": dispatch_quantity,
    }


def resolve_report_date(report_date):
    if report_date:
        return getdate(report_date)

    latest_production_date = frappe.db.sql(
        """
        select max(ifnull(se.posting_date, date(bpr.modified))) as report_date
        from `tabBatch Production Record` bpr
        left join `tabStock Entry` se on se.name = bpr.stock_entry
        where bpr.docstatus = 1
        """,
        as_dict=True,
    )
    if latest_production_date and latest_production_date[0].report_date:
        return getdate(latest_production_date[0].report_date)

    return getdate(nowdate())


def get_daily_production_by_fg(report_date) -> list[dict[str, object]]:
    return frappe.db.sql(
        """
        select
            bpr.item_code,
            coalesce(i.item_name, bpr.item_code) as item_name,
            round(sum(bpr.produced_qty), 3) as qty
        from `tabBatch Production Record` bpr
        left join `tabStock Entry` se on se.name = bpr.stock_entry
        left join `tabItem` i on i.name = bpr.item_code
        where bpr.docstatus = 1
          and ifnull(se.posting_date, date(bpr.modified)) = %(report_date)s
        group by bpr.item_code, item_name
        order by qty desc, bpr.item_code asc
        limit 10
        """,
        {"report_date": report_date},
        as_dict=True,
    )


def get_extruder_utilization(report_date) -> list[dict[str, object]]:
    job_card_columns = set(frappe.db.get_table_columns("Job Card") or [])
    if "workstation" not in job_card_columns:
        return []

    time_fields = [field for field in ("total_time_in_mins", "time_required") if field in job_card_columns]
    if not time_fields:
        return []

    machine_names = {machine["name"] for machine in PLANT_MACHINES}
    machine_aliases = {machine["alias"] for machine in PLANT_MACHINES}
    time_expr = "coalesce(" + ", ".join(f"jc.{field}" for field in time_fields) + ", 0)"
    date_field = "jc.posting_date" if "posting_date" in job_card_columns else "date(jc.modified)"

    rows = frappe.db.sql(
        f"""
        select
            coalesce(ws.workstation_name, jc.workstation) as workstation,
            round(sum({time_expr}), 2) as runtime_minutes
        from `tabJob Card` jc
        left join `tabWorkstation` ws on ws.name = jc.workstation
        where jc.docstatus < 2
          and {date_field} = %(report_date)s
        group by jc.workstation, workstation
        order by workstation asc
        """,
        {"report_date": report_date},
        as_dict=True,
    )

    available_minutes = 24 * 60
    output = []
    for row in rows:
        workstation_name = row.workstation
        if workstation_name not in machine_names and workstation_name not in machine_aliases:
            continue
        runtime_minutes = flt(row.runtime_minutes)
        output.append(
            {
                "workstation": workstation_name,
                "runtime_minutes": runtime_minutes,
                "available_minutes": available_minutes,
                "utilization_pct": round(min((runtime_minutes / available_minutes) * 100, 100), 2),
            }
        )
    return output


def get_rm_consumption(report_date) -> list[dict[str, object]]:
    return frappe.db.sql(
        """
        select
            sed.item_code,
            coalesce(i.item_name, sed.item_code) as item_name,
            round(sum(sed.qty), 3) as qty
        from `tabStock Entry Detail` sed
        inner join `tabStock Entry` se on se.name = sed.parent
        left join `tabItem` i on i.name = sed.item_code
        where se.docstatus = 1
          and ifnull(se.stock_entry_type, se.purpose) = 'Manufacture'
          and ifnull(sed.is_finished_item, 0) = 0
          and se.posting_date = %(report_date)s
        group by sed.item_code, item_name
        order by qty desc, sed.item_code asc
        limit 10
        """,
        {"report_date": report_date},
        as_dict=True,
    )


def get_qc_failures(report_date) -> list[dict[str, object]]:
    output = OrderedDict(
        {
            "RM Hold": 0,
            "RM Rejected": 0,
            "Final QC Hold": 0,
            "Final QC Rejected": 0,
            "Quality Inspection Rejected": 0,
        }
    )

    rm_rows = frappe.db.sql(
        """
        select decision, count(*) as count
        from `tabRM QC Decision`
        where docstatus = 1
          and decision in ('Hold', 'Rejected')
          and date(modified) = %(report_date)s
        group by decision
        """,
        {"report_date": report_date},
        as_dict=True,
    )
    for row in rm_rows:
        output[f"RM {row.decision}"] = int(row.count)

    final_rows = frappe.db.sql(
        """
        select status, count(*) as count
        from `tabFinal QC Release`
        where docstatus = 1
          and status in ('Hold', 'Rejected')
          and date(ifnull(released_on, modified)) = %(report_date)s
        group by status
        """,
        {"report_date": report_date},
        as_dict=True,
    )
    for row in final_rows:
        output[f"Final QC {row.status}"] = int(row.count)

    inspection_count = frappe.db.count(
        "Quality Inspection",
        filters={
            "docstatus": 1,
            "status": "Rejected",
            "modified": ("between", [f"{report_date} 00:00:00", f"{report_date} 23:59:59"]),
        },
    )
    output["Quality Inspection Rejected"] = int(inspection_count)

    return [{"label": label, "count": count} for label, count in output.items()]


def get_dispatch_quantity(report_date) -> list[dict[str, object]]:
    return frappe.db.sql(
        """
        select
            dni.item_code,
            coalesce(i.item_name, dni.item_code) as item_name,
            round(sum(dni.qty), 3) as qty
        from `tabDelivery Note Item` dni
        inner join `tabDelivery Note` dn on dn.name = dni.parent
        left join `tabItem` i on i.name = dni.item_code
        where dn.docstatus = 1
          and dn.posting_date = %(report_date)s
        group by dni.item_code, item_name
        order by qty desc, dni.item_code asc
        limit 10
        """,
        {"report_date": report_date},
        as_dict=True,
    )


def get_cost_by_product(report_date) -> list[dict[str, object]]:
    return frappe.db.sql(
        """
        select
            prod.item_code,
            prod.item_name,
            round(sum(prod.produced_qty), 3) as produced_qty,
            round(sum(coalesce(rm.rm_cost_total, 0)), 2) as rm_cost_total,
            round(
                case when sum(prod.produced_qty) > 0
                    then sum(coalesce(rm.rm_cost_total, 0)) / sum(prod.produced_qty)
                    else 0
                end,
                4
            ) as rm_cost_per_kg,
            round(sum(coalesce(fg.fg_cost_total, coalesce(rm.rm_cost_total, 0))), 2) as manufacturing_cost_total,
            round(
                case when sum(prod.produced_qty) > 0
                    then sum(coalesce(fg.fg_cost_total, coalesce(rm.rm_cost_total, 0))) / sum(prod.produced_qty)
                    else 0
                end,
                4
            ) as manufacturing_cost_per_kg
        from (
            select
                bpr.stock_entry,
                bpr.item_code,
                coalesce(i.item_name, bpr.item_code) as item_name,
                bpr.produced_qty
            from `tabBatch Production Record` bpr
            left join `tabStock Entry` se on se.name = bpr.stock_entry
            left join `tabItem` i on i.name = bpr.item_code
            where bpr.docstatus = 1
              and ifnull(se.posting_date, date(bpr.modified)) = %(report_date)s
        ) prod
        left join (
            select
                sed.parent as stock_entry,
                sum(
                    abs(
                        coalesce(
                            sed.basic_amount,
                            sed.amount,
                            abs(sed.qty) * coalesce(sed.basic_rate, sed.valuation_rate, i.valuation_rate, 0)
                        )
                    )
                ) as rm_cost_total
            from `tabStock Entry Detail` sed
            inner join `tabStock Entry` se on se.name = sed.parent
            left join `tabItem` i on i.name = sed.item_code
            where se.docstatus = 1
              and ifnull(se.stock_entry_type, se.purpose) = 'Manufacture'
              and ifnull(sed.is_finished_item, 0) = 0
              and se.posting_date = %(report_date)s
            group by sed.parent
        ) rm on rm.stock_entry = prod.stock_entry
        left join (
            select
                sed.parent as stock_entry,
                sum(
                    abs(
                        coalesce(
                            sed.basic_amount,
                            sed.amount,
                            abs(sed.qty) * coalesce(sed.basic_rate, sed.valuation_rate, i.valuation_rate, 0)
                        )
                    )
                ) as fg_cost_total
            from `tabStock Entry Detail` sed
            inner join `tabStock Entry` se on se.name = sed.parent
            left join `tabItem` i on i.name = sed.item_code
            where se.docstatus = 1
              and ifnull(se.stock_entry_type, se.purpose) = 'Manufacture'
              and ifnull(sed.is_finished_item, 0) = 1
              and se.posting_date = %(report_date)s
            group by sed.parent
        ) fg on fg.stock_entry = prod.stock_entry
        group by prod.item_code, prod.item_name
        order by produced_qty desc, prod.item_code asc
        """,
        {"report_date": report_date},
        as_dict=True,
    )


def get_profit_by_product(report_date, cost_by_product: list[dict[str, object]]) -> list[dict[str, object]]:
    if not cost_by_product:
        return []

    cost_map = {row["item_code"]: row for row in cost_by_product}
    dispatch_rows = frappe.db.sql(
        """
        select
            dni.item_code,
            coalesce(i.item_name, dni.item_code) as item_name,
            round(sum(dni.qty), 3) as dispatched_qty,
            round(sum(coalesce(dni.base_amount, dni.amount, dni.qty * dni.rate)), 2) as sales_amount_total,
            round(
                case when sum(dni.qty) > 0
                    then sum(coalesce(dni.base_amount, dni.amount, dni.qty * dni.rate)) / sum(dni.qty)
                    else 0
                end,
                4
            ) as selling_price_per_kg
        from `tabDelivery Note Item` dni
        inner join `tabDelivery Note` dn on dn.name = dni.parent
        left join `tabItem` i on i.name = dni.item_code
        where dn.docstatus = 1
          and dn.posting_date = %(report_date)s
        group by dni.item_code, item_name
        order by dispatched_qty desc, dni.item_code asc
        """,
        {"report_date": report_date},
        as_dict=True,
    )

    output = []
    for row in dispatch_rows:
        cost_row = cost_map.get(row["item_code"])
        if not cost_row:
            continue

        manufacturing_cost_per_kg = flt(cost_row.get("manufacturing_cost_per_kg"))
        dispatched_qty = flt(row.get("dispatched_qty"))
        selling_price_per_kg = flt(row.get("selling_price_per_kg"))
        gross_profit_per_kg = round(selling_price_per_kg - manufacturing_cost_per_kg, 4)
        gross_profit_total = round(gross_profit_per_kg * dispatched_qty, 2)

        output.append(
            {
                "item_code": row["item_code"],
                "item_name": row["item_name"],
                "dispatched_qty": dispatched_qty,
                "sales_amount_total": flt(row.get("sales_amount_total")),
                "selling_price_per_kg": selling_price_per_kg,
                "manufacturing_cost_per_kg": manufacturing_cost_per_kg,
                "gross_profit_per_kg": gross_profit_per_kg,
                "gross_profit_total": gross_profit_total,
            }
        )

    output.sort(key=lambda row: (-row["gross_profit_total"], row["item_code"]))
    return output

File: calco_production/page/plant_production_dashboard/plant_production_dashboard.js
frappe.pages["plant-production-dashboard"].on_page_load = function (wrapper) {
  new PlantProductionDashboard(wrapper);
};

class PlantProductionDashboard {
  constructor(wrapper) {
    this.wrapper = $(wrapper);
    this.charts = {};
    this.make_page();
    this.make_layout();
    this.refresh();
  }

  make_page() {
    this.page = frappe.ui.make_app_page({
      parent: this.wrapper,
      title: "Plant Production Dashboard",
      single_column: true,
    });

    this.date_field = this.page.add_field({
      label: "Date",
      fieldname: "report_date",
      fieldtype: "Date",
      default: "",
      change: () => this.refresh(),
    });

    this.page.set_primary_action("Refresh", () => this.refresh(), "refresh");
  }

  make_layout() {
    this.page.main.html(`
      <div class="plant-dashboard">
        <div class="plant-dashboard__cards"></div>
        <div class="plant-dashboard__grid">
          <section class="plant-dashboard__panel">
            <h4>Daily Production by FG</h4>
            <div class="plant-dashboard__chart" data-chart="production"></div>
          </section>
          <section class="plant-dashboard__panel">
            <h4>Output per Machine</h4>
            <div class="plant-dashboard__chart" data-chart="machine-output"></div>
          </section>
          <section class="plant-dashboard__panel">
            <h4>Output per Operator</h4>
            <div class="plant-dashboard__chart" data-chart="operator-output"></div>
          </section>
          <section class="plant-dashboard__panel">
            <h4>Output per Shift</h4>
            <div class="plant-dashboard__chart" data-chart="shift-output"></div>
          </section>
          <section class="plant-dashboard__panel">
            <h4>Extruder Utilization</h4>
            <div class="plant-dashboard__chart" data-chart="utilization"></div>
          </section>
          <section class="plant-dashboard__panel">
            <h4>RM Consumption</h4>
            <div class="plant-dashboard__chart" data-chart="rm"></div>
          </section>
          <section class="plant-dashboard__panel">
            <h4>RM Cost per Kg</h4>
            <div class="plant-dashboard__chart" data-chart="rm-cost"></div>
          </section>
          <section class="plant-dashboard__panel">
            <h4>Manufacturing Cost per Kg</h4>
            <div class="plant-dashboard__chart" data-chart="manufacturing-cost"></div>
          </section>
          <section class="plant-dashboard__panel">
            <h4>Profit per Product</h4>
            <div class="plant-dashboard__chart" data-chart="profit"></div>
          </section>
          <section class="plant-dashboard__panel">
            <h4>QC Failures</h4>
            <div class="plant-dashboard__chart" data-chart="qc"></div>
          </section>
          <section class="plant-dashboard__panel">
            <h4>Dispatch Quantity</h4>
            <div class="plant-dashboard__chart" data-chart="dispatch"></div>
          </section>
        </div>
      </div>
    `);

    if (!document.getElementById("plant-production-dashboard-style")) {
      const style = document.createElement("style");
      style.id = "plant-production-dashboard-style";
      style.textContent = `
        .plant-dashboard__cards {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(180px, 1fr));
          gap: 16px;
          margin-bottom: 20px;
        }
        .plant-dashboard__card {
          border: 1px solid var(--border-color);
          border-radius: 12px;
          background: var(--fg-color);
          padding: 16px;
        }
        .plant-dashboard__card-label {
          color: var(--text-muted);
          font-size: 12px;
          margin-bottom: 6px;
          text-transform: uppercase;
          letter-spacing: 0.04em;
        }
        .plant-dashboard__card-value {
          font-size: 24px;
          font-weight: 700;
        }
        .plant-dashboard__grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(420px, 1fr));
          gap: 16px;
        }
        .plant-dashboard__panel {
          border: 1px solid var(--border-color);
          border-radius: 12px;
          background: var(--fg-color);
          padding: 16px;
          min-height: 340px;
        }
        .plant-dashboard__panel h4 {
          margin: 0 0 12px;
          font-size: 16px;
        }
        .plant-dashboard__chart {
          min-height: 260px;
        }
        .plant-dashboard__empty {
          color: var(--text-muted);
          padding: 24px 0;
        }
      `;
      document.head.appendChild(style);
    }
  }

  refresh() {
    frappe.call({
      method:
        "calco_erp.calco_production.page.plant_production_dashboard.plant_production_dashboard.get_dashboard_data",
      args: this.date_field.get_value()
        ? {
            report_date: this.date_field.get_value(),
          }
        : {},
      freeze: true,
      freeze_message: "Loading plant dashboard...",
    }).then((r) => {
      this.data = r.message || {};
      if (this.data.report_date && this.date_field.get_value() !== this.data.report_date) {
        this.date_field.set_value(this.data.report_date);
      }
      this.render_cards();
      this.render_charts();
    });
  }

  render_cards() {
    const cards = this.data.cards || {};
    const formatQty = (value) => format_number(value || 0, null, 3);
    const toInt = (value) => Number.parseInt(value || 0, 10) || 0;
    const items = [
      { label: "Production Qty", value: `${formatQty(cards.production_qty)} Kg` },
      { label: "Dispatch Qty", value: `${formatQty(cards.dispatch_qty)} Kg` },
      { label: "RM Consumed", value: `${formatQty(cards.rm_consumption_qty)} Kg` },
      { label: "QC Failures", value: toInt(cards.qc_failure_count || 0) },
      { label: "Machines Active", value: toInt(cards.machine_count || 0) },
      { label: "Operators Active", value: toInt(cards.operator_count || 0) },
      { label: "Shifts Active", value: toInt(cards.shift_count || 0) },
      { label: "Avg RM Cost / Kg", value: `Rs ${formatQty(cards.avg_rm_cost_per_kg)}` },
      { label: "Avg Mfg Cost / Kg", value: `Rs ${formatQty(cards.avg_manufacturing_cost_per_kg)}` },
      { label: "Gross Profit", value: `Rs ${formatQty(cards.gross_profit_total)}` },
      { label: "Avg Extruder Utilization", value: `${formatQty(cards.avg_extruder_utilization)}%` },
    ];

    const html = items
      .map(
        (item) => `
          <div class="plant-dashboard__card">
            <div class="plant-dashboard__card-label">${frappe.utils.escape_html(item.label)}</div>
            <div class="plant-dashboard__card-value">${frappe.utils.escape_html(String(item.value))}</div>
          </div>
        `
      )
      .join("");

    this.page.main.find(".plant-dashboard__cards").html(html);
  }

  render_charts() {
    this.renderBarChart("production", this.data.daily_production_by_fg || [], "qty", ["#c62828"]);
    this.renderBarChart("machine-output", this.data.output_by_machine || [], "produced_qty", ["#ef6c00"], "machine");
    this.renderBarChart(
      "operator-output",
      this.data.output_by_operator || [],
      "produced_qty",
      ["#00897b"],
      "operator_name"
    );
    this.renderBarChart("shift-output", this.data.output_by_shift || [], "produced_qty", ["#5e35b1"], "shift_type");
    this.renderBarChart("utilization", this.data.extruder_utilization || [], "utilization_pct", ["#1565c0"], "workstation", "%");
    this.renderBarChart("rm", this.data.rm_consumption || [], "qty", ["#2e7d32"]);
    this.renderBarChart("rm-cost", this.data.cost_by_product || [], "rm_cost_per_kg", ["#00838f"], "item_code", " Rs/Kg");
    this.renderBarChart(
      "manufacturing-cost",
      this.data.cost_by_product || [],
      "manufacturing_cost_per_kg",
      ["#ad1457"],
      "item_code",
      " Rs/Kg"
    );
    this.renderBarChart("profit", this.data.profit_by_product || [], "gross_profit_total", ["#2e7d32"], "item_code", " Rs");
    this.renderDonutChart("qc", this.data.qc_failures || [], "count", ["#ef5350", "#ffa726", "#29b6f6", "#ab47bc", "#66bb6a"], "label");
    this.renderBarChart("dispatch", this.data.dispatch_quantity || [], "qty", ["#6a1b9a"]);
  }

  renderBarChart(chartKey, rows, valueKey, colors, labelKey = "item_code", suffix = " Kg") {
    const target = this.page.main.find(`[data-chart="${chartKey}"]`);
    if (!rows.length) {
      target.html('<div class="plant-dashboard__empty">No data for the selected date.</div>');
      return;
    }

    target.empty();
    this.destroyChart(chartKey);
    this.charts[chartKey] = new frappe.Chart(target[0], {
      type: "bar",
      height: 260,
      colors,
      data: {
        labels: rows.map((row) => row[labelKey]),
        datasets: [
          {
            values: rows.map((row) => row[valueKey] || 0),
          },
        ],
      },
      axisOptions: {
        xAxisMode: "tick",
        yAxisMode: "span",
        xIsSeries: 1,
      },
      tooltipOptions: {
        formatTooltipY: (value) => `${format_number(value || 0, null, 3)}${suffix}`,
      },
    });
  }

  renderDonutChart(chartKey, rows, valueKey, colors, labelKey = "label") {
    const target = this.page.main.find(`[data-chart="${chartKey}"]`);
    const filteredRows = rows.filter((row) => (Number(row[valueKey] || 0) || 0) > 0);
    if (!filteredRows.length) {
      target.html('<div class="plant-dashboard__empty">No QC failures for the selected date.</div>');
      return;
    }

    target.empty();
    this.destroyChart(chartKey);
    this.charts[chartKey] = new frappe.Chart(target[0], {
      type: "donut",
      height: 260,
      colors,
      data: {
        labels: filteredRows.map((row) => row[labelKey]),
        datasets: [
          {
            values: filteredRows.map((row) => row[valueKey] || 0),
          },
        ],
      },
    });
  }

  destroyChart(chartKey) {
    if (this.charts[chartKey]) {
      this.charts[chartKey].destroy();
      delete this.charts[chartKey];
    }
  }
}
